# cite
## checkm
[GitHub - chklovski/CheckM2: Assessing the quality of metagenome-derived genome bins using machine learning](https://github.com/chklovski/CheckM2)
## checkm2
[GitHub - chklovski/CheckM2: Assessing the quality of metagenome-derived genome bins using machine learning](https://github.com/chklovski/CheckM2)

# bash
## checkm
```
cd /number1/SUSTech_project/workflow_v2/stage_IV_prokka/
cd I_genome_pre/I_I-checkm/
conda activate checkm 
mkdir checkm 
cd checkm
checkm taxon_set genus   Prochlorococcus   Prochlorococcus.ms
checkm analyze Prochlorococcus.ms /number1/SUSTech_project/workflow_v2/stage_IV_prokka/genome_data_checkmdefault_drep5098/genome_short_contig/ Prochlorococcus_result -x fasta  -t 40

mkdir qa
for num in 1 2 3 4 5 6 7 8 9
do
checkm qa Prochlorococcus.ms Prochlorococcus_result -o $num --tab_table -f qa/qa_$num.txt
done


```
## checkm2
```
cd /number1/SUSTech_project/workflow_v2/stage_IV_prokka/
conda activate checkm2
cd I_genome_pre/I_I-checkm/

mkdir checkm2
cd checkm2

checkm2 predict --threads 30 -i /number1/SUSTech_project/workflow_v2/stage_IV_prokka/genome_data_checkmdefault_drep5098/genome_short_contig/   -o Prochlorococcus_result_checkm2 --allmodels -x fasta --force -t 40 

```


# snakemake
## checkm

## checkm2
```


```
